cloud/amazon/ec2_vpc_nat_gateway.py

```diff
@@ -0,0 +1,356 @@
+                                               AllocationId=module.params.get('allocation_id'))
+        wait = module.params.get('wait')
+        start_time = datetime.datetime.utcnow()
+        results = dict(changed=True)
+        if wait:
+            wait_timeout = time.time() + module.params.get('wait_timeout')
+            while wait_timeout > time.time():
+                if 'nat_gateway' in results:
+                    break
+                for present_status in get_nat_gateway_status_list(ec2_client=ec2_client, module=module):
+                    create_time = present_status.get('created')
+                    if present_status.get('state') == 'failed' and create_time.replace(tzinfo=None) >= start_time:
+                        module.fail_json(msg="Failed to create VPC NAT Gateway")
+                    elif present_status.get('state') == 'available':
+                        results['nat_gateway'] = camel_dict_to_snake_dict(result['NatGateway'])
+                        break
+                    else:
+                        time.sleep(5)
+            else:
+                module.fail_json(msg="Waited too long for VPC NAT Gateway to be created.")
+        module.exit_json(**results)
+    except botocore.exceptions.ClientError as ce:
+        module.fail_json(msg=ce.message)
+
+
+def ensure_nat_gateway_absent(ec2_client=None, module=None):
+    """ Ensure the specified NAT gateway does not exist and call delete if it does """
+    try:
+        results = dict(changed=False)
+        nat_gateways_result = ec2_client.describe_nat_gateways(NatGatewayIds=[module.params.get('nat_gateway_id')])
+        if nat_gateways_result:
+            nat_gateway = nat_gateways_result.get('NatGateways')[0]
+            if nat_gateway.get('State') in ('pending',
+                                            'available'):
+                if module.check_mode:
+                    results['changed'] = True
+                    module.exit_json(**results)
+                else:
+                    delete_nat_gateway(ec2_client=ec2_client,
+                                       module=module,
+                                       nat_gateway=nat_gateway)
+        module.exit_json(**results)
+    except botocore.exceptions.ClientError as ce:
+        if ce.response['Error']['Code'] == "NatGatewayMalformed":
+            module.fail_json(msg="Invalid NAT Gateway ID")
+        else:
+            module.fail_json(msg=ce.message)
+
+
+def ensure_nat_gateway_present(ec2_client=None, module=None):
+    """ Ensure NAT gateway with specified parameters exists and call create if it does not """
+    # Check allocation-id refers to existing EIP allocation
+    allocation_id = module.params.get('allocation_id')
+    try:
+        ec2_client.describe_addresses(AllocationIds=[module.params.get('allocation_id')])
+    except botocore.exceptions.ClientError:
+        module.fail_json(msg="allocation: %s does not exist." % allocation_id)
```

Comment:
Probably not wise to assume this ClientError is because of allocation not existing.  Can you check the response code?

Also as per the others...

Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`